Support unicode identifiers in comments

314eter opened the pull request:
The comment lexer was not updated to support unicode identifiers in #12664.

If you try to comment out this block of code, the comment is not valid:
```ocaml
(* let é' = ignore in é' '"' *)
```

This can be fixed by using `ident_ext` instead of `ident` in the comment lexer.

I don't think we need to validate with `Utf8_lexeme.validate_identifier`. This means that the same comment with an invalid identifier will be accepted too:
```ocaml
(* let 🐫' = ignore in 🐫' '"' *)
```
Avoiding that will be complicated, but accepting invalid code in comments is less problematic than not accepting valid code.

314eter committed:
Support unicode identifiers in comments

Octachron commented:
Good catch ! Would you mind adding your example in the `testsuite/tests/parsing/latin9.ml` test?
I agree that it is not necessary to reject invalid identifiers in comments.

gasche commented:
Unless I am missing something, this was the last use of `ident` as a lexing symbol, and now all uses of `ident` have been replaced by `ident_ext`. I think we should remove `ident` to avoid this class of mistakes.

314eter committed:
Add test for unicode identifier in comment

314eter commented:
I added the test in `testsuite/tests/parsing/comments.ml`, since there was already a test for commented code there.

I already removed `ident`, it was indeed the last use.

I'm not sure what to do with the Changes file. Should this be in 5.3.0?

Octachron commented:
5.3.0 has been out since the end of this week, thus you can add the changes entry in the bug subsection of  "Working version" section.

nojb commented:
> 5.3.0 has been out since the end of this week, thus you can add the changes entry in the bug subsection of "Working version" section.

Sorry, just to confirm (I'm catching up after holidays), here "has been out since" should be "will be out by"?

Octachron commented:
It was a future past tense indeed, we are not yet Friday. I am planning the release for Wednesday, and only critical bug fixes will interrupt the release.

hannesm commented:
I personally would love to have this in 5.3 for consistency reasons - as I understand the unicode identifiers have been introduced in 5.3, and it would be sad to get errors from commenting out the code. I also think this is a pretty low-risk PR.

Octachron commented:
Note that the bug requires a non-ascii identifier ending with `'` followed by one character, followed by `'"'` to trigger.
The bug is thus quite specific and would not impact most code with unicode identifiers.

314eter commented:
The chances of it happening in real code are indeed low. I only discovered it because I was updating [tree-sitter-ocaml](https://github.com/tree-sitter/tree-sitter-ocaml) with the unicode changes.

314eter committed:
Changes
